packages/react-query/src/__tests__/suspense.test.tsx

```diff
@@ -0,0 +1,231 @@
+  })
+
+  it('should respect staleTime when value is greater than 1000ms', async () => {
+    const key = queryKey()
+    const staleTime = 2000
+    let fetchCount = 0
+
+    function Component() {
+      const result = useSuspenseQuery({
+        queryKey: key,
+        queryFn: async () => {
+          fetchCount++
+          await sleep(10)
+          return 'data'
+        },
+        staleTime,
+      })
+      return <div>data: {result.data}</div>
+    }
+
+    const rendered = renderWithClient(
+      queryClient,
+      <Suspense fallback="loading">
+        <Component />
+      </Suspense>,
+    )
+
+    await waitFor(() => rendered.getByText('data: data'))
+
+    // Manually trigger a remount
+    rendered.rerender(
+      <Suspense fallback="loading">
+        <Component />
+      </Suspense>,
+    )
+
+    // Wait longer than 1000ms but less than staleTime
+    await act(async () => {
+      await sleep(1500)
+    })
+
+    // Should not refetch as we're still within the staleTime
+    expect(fetchCount).toBe(1)
+  })
+
+  it('should enforce minimum staleTime when undefined is provided', async () => {
+    const key = queryKey()
+    let fetchCount = 0
+
+    function Component() {
+      const result = useSuspenseQuery({
+        queryKey: key,
+        queryFn: async () => {
+          fetchCount++
+          await sleep(10)
+          return 'data'
+        },
+        // Explicitly set staleTime as undefined
+        staleTime: undefined,
+      })
+      return <div>data: {result.data}</div>
+    }
+
+    const rendered = renderWithClient(
+      queryClient,
+      <Suspense fallback="loading">
+        <Component />
+      </Suspense>,
+    )
+
+    await waitFor(() => rendered.getByText('data: data'))
+
+    rendered.rerender(
+      <Suspense fallback="loading">
+        <Component />
+      </Suspense>,
+    )
+
+    // Wait less than enforced 1000ms
+    await act(async () => {
+      await sleep(500)
+    })
+
+    // Should not refetch as minimum staleTime of 1000ms is enforced
+    expect(fetchCount).toBe(1)
+  })
+
+  it('should respect staleTime when function returns value greater than 1000ms', async () => {
+    const key = queryKey()
+    let fetchCount = 0
+
+    function Component() {
+      const result = useSuspenseQuery({
+        queryKey: key,
+        queryFn: async () => {
+          fetchCount++
+          await sleep(10)
+          return 'data'
+        },
+        staleTime: () => 3000,
+      })
+      return <div>data: {result.data}</div>
+    }
+
+    const rendered = renderWithClient(
+      queryClient,
+      <Suspense fallback="loading">
+        <Component />
+      </Suspense>,
+    )
+
+    await waitFor(() => rendered.getByText('data: data'))
+
+    rendered.rerender(
+      <Suspense fallback="loading">
+        <Component />
+      </Suspense>,
+    )
+
+    // Wait longer than 1000ms but less than returned staleTime
+    await act(async () => {
+      await sleep(2000)
+    })
```

Comment:
this will make for a very long test. `fakeTimers` from vitest are probably the way to go here. We have some examples already, e.g. here:

https://github.com/TanStack/query/blob/1eebfc9d3355c411b16146cc8df76c18e63d59ad/packages/react-query/src/__tests__/useSuspenseQueries.test.tsx#L37

Reply:
<img width="518" alt="image" src="https://github.com/user-attachments/assets/b3f5b600-cb82-4243-bb8e-24fbf7aff397" />

I improved the test code through a lot of thought to improve the test time, and the useFakeTimers you gave me were especially helpful. Thanks.

46057ca7076a40c7019346f7d0e4720d00f078f3